internal/privatemessaging/privatemessaging.go

```diff
@@ -190,39 +197,71 @@ func (pm *privateMessaging) dispatchBatchCommon(ctx context.Context, state *batc
 	return pm.sendData(ctx, tw, nodes)
 }
 
-func (pm *privateMessaging) transferBlobs(ctx context.Context, data fftypes.DataArray, txid *fftypes.UUID, node *fftypes.Identity) error {
-	// Send all the blobs associated with this batch
-	for _, d := range data {
-		if d.Blob != nil {
-			if d.Blob.Hash == nil {
-				return i18n.NewError(ctx, i18n.MsgDataMissingBlobHash, d.ID)
-			}
-
-			blob, err := pm.database.GetBlobMatchingHash(ctx, d.Blob.Hash)
-			if err != nil {
-				return err
-			}
-			if blob == nil {
-				return i18n.NewError(ctx, i18n.MsgBlobNotFound, d.Blob)
-			}
-
-			op := fftypes.NewOperation(
-				pm.exchange,
-				d.Namespace,
-				txid,
-				fftypes.OpTypeDataExchangeSendBlob)
-			addTransferBlobInputs(op, node.ID, blob.Hash)
-			log.L(ctx).Debugf("Transferring blob %s for data %s in operation %s", d.Blob.Hash, d.ID, op.ID)
-			if err = pm.operations.AddOrReuseOperation(ctx, op); err != nil {
-				return err
+func (pm *privateMessaging) prepareBlobTransfers(ctx context.Context, data fftypes.DataArray, txid *fftypes.UUID, node *fftypes.Identity) ([]*blobTransferTracker, error) {
+
+	operations := make([]*blobTransferTracker, 0)
+
+	// Build all the operations needed to send the blobs in a single DB transaction
+	err := pm.database.RunAsGroup(ctx, func(ctx context.Context) error {
+		for _, d := range data {
+			if d.Blob != nil {
+				if d.Blob.Hash == nil {
+					return i18n.NewError(ctx, i18n.MsgDataMissingBlobHash, d.ID)
+				}
+
+				blob, err := pm.database.GetBlobMatchingHash(ctx, d.Blob.Hash)
+				if err != nil {
+					return err
+				}
+				if blob == nil {
+					return i18n.NewError(ctx, i18n.MsgBlobNotFound, d.Blob)
+				}
+
+				op := fftypes.NewOperation(
+					pm.exchange,
+					d.Namespace,
+					txid,
+					fftypes.OpTypeDataExchangeSendBlob)
+				addTransferBlobInputs(op, node.ID, blob.Hash)
+				if err = pm.operations.AddOrReuseOperation(ctx, op); err != nil {
+					return err
+				}
+
+				// Check the operation isn't already successful from a previous run
```

Comment:
Is there a check for previous success..?

Reply:
No :)